### PR TITLE
Parameterized FIRE (Adding Options for FIRE)

### PR DESCRIPTION
**Pull request by Mars-Cat2023**
Up to this moment, you can find these following **successful** changes in this PR:

1. v3 - **Adding one argument**: `–fire_log_bias` (used in calculating log_rel_distance and log_pos_normalizer)
2. v4 - **About new hyperparameters**:
          Based on original interface, "def __init__(self, config, num_heads=12, mlp_width=32, init_c=0.1, init_L=512.0, eps=1e-6):"
where
     - (1) **num_heads** is not changeable, because in our implementation, we always have **num_heads = n_head**.

    - (2) **eps** is not worth parameterizing (less significant to the performance of our whole model theoretically).

    Therefore, I **parameterized** `mlp_width=32, init_c=0.1, init_L=512.0.` as `--fire_mlp_width, --fire_init_c, --fire_init_L`, respectively.
    Additionally, I **add two more options** based on their ablation studies:
    - (1) `--fire_num_hidden_layers` **(see Appendix B-2 of FIRE paper)** to represent the number of hidden layers (sigmas) in mlp in FIRE **without counting outermost sigma**.

          Specially, 
          "fire_num_hidden_layers = 0" means no hidden layers: 
                    nn.Linear(input_dim, output_dim).
          
          "fire_num_hidden_layers = 1" (default) means one hidden layer:
                = The one implemented by the provided code in paper.
                    nn.Linear(input_dim, mlp_width), RELU, nn.Linear(mlp_width, output_dim)
          
          "fire_num_hidden_layers = 2" means two hidden layers:
                = The best choice that author mentioned in main text body of paper, but not implemented in the provided code.
                    nn.Linear(input_dim, mlp_width), RELU, nn.Linear(mlp_width, mlp_width), RELU, nn.Linear(mlp_width, output_dim)
          
          "fire_num_hidden_layers = 3, 4, 5, 6..." are also supported here. 
           
          [Attention]: As I stressed in our group meeting, this is another difference between the text \ 
                     and the provided code in FIRE paper, where the number of hidden layers is not the same as each other.

   - (2) `--fire_outermost_sigma` **(see Appendix B-4 of FIRE paper)** to change whether or not applying addition final activation to the MLP final output in FIRE.

### CODE REVIEW

**Comment by Mars-Cat2023** [comment]
"v4" has been verified locally:
**{shakespeare_char, softmax, use_fire_embeddings} with all default FIRE configuration** will give you the same result as the code before parameterizing FIRE:
image

**Comment by Mars-Cat2023** [comment]
Up to v4, It's Ready to Merge !! 😊🤩😁 @gkielian
